src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx

```diff
@@ -0,0 +1,208 @@
+import { Button, Flex, List, useColorModeValue } from '@chakra-ui/react'
+import { DefiAction } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import { useCallback, useMemo } from 'react'
+import { useTranslate } from 'react-polyglot'
+import { Amount } from 'components/Amount/Amount'
+import { AssetCell } from 'components/StakingVaults/Cells'
+import { RawText } from 'components/Text'
+import { bn, bnOrZero } from 'lib/bignumber/bignumber'
+import type {
+  LpEarnOpportunityType,
+  StakingEarnOpportunityType,
+} from 'state/slices/opportunitiesSlice/types'
+import {
+  getRewardBalances,
+  getUnderlyingAssetIdsBalances,
+} from 'state/slices/opportunitiesSlice/utils'
+import {
+  selectAssetById,
+  selectAssets,
+  selectMarketDataSortedByMarketCap,
+} from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+import { NestedAsset } from './NestedAsset'
+import { OpportunityRowGrid } from './OpportunityTableHeader'
+
+type OpportunityRowProps<T extends StakingEarnOpportunityType | LpEarnOpportunityType> = {
+  onClick: (opportunity: T, action: DefiAction) => void
```

Comment:
Usually not a fan of passing an opportunity as an object by reference instead of an `opportunityId`, but happy here to make the generic work